src/cogent3/core/new_sequence.py

```diff
@@ -0,0 +1,2413 @@
+                new[new < 0] = 0
+            elif coord.min() < len(self) < coord.max():
+                new[new > len(self)] = len(self)
+            elif coord[0] == coord[1] or coord.min() > len(self) or coord.max() < 0:
+                # would really to adjust pre and post to be up to the next span
+                continue
+            new_spans.append(new.tolist())
+
+        fmap = FeatureMap.from_locations(locations=new_spans, parent_length=len(self))
+        if pre or post:
+            # create a lost span to represent the segment missing from
+            # the instance
+            spans = list(fmap.spans)
+            if pre:
+                spans.insert(0, LostSpan(pre))
+            if post:
+                spans.append(LostSpan(post))
+            fmap = FeatureMap(spans=spans, parent_length=len(self))
+
+        if seq_rced:
+            fmap = fmap.nucleic_reversed()
+
+        feature.pop("on_alignment", None)
+        feature.pop("seqid", None)
+        return Feature(parent=self, seqid=self.name, map=fmap, **feature)
+
+    def annotate_from_gff(self, f: os.PathLike, offset=None):
+        """copies annotations from a gff file to self,
+
+        Parameters
+        ----------
+        f : path to gff annotation file.
+        offset : Optional, the offset between annotation coordinates and sequence coordinates.
+        """
+        if isinstance(self.annotation_db, GenbankAnnotationDb):
+            raise ValueError("GenbankAnnotationDb already attached")
+
+        self.annotation_db = load_annotations(
+            path=f, seqids=self.name, db=self.annotation_db
+        )
+
+        if offset:
+            self.annotation_offset = offset
+
+    def add_feature(
+        self,
+        *,
+        biotype: str,
+        name: str,
+        spans: list[tuple[int, int]],
+        parent_id: OptStr = None,
+        strand: OptStr = None,
+        on_alignment: bool = False,
+        seqid: OptStr = None,
+    ) -> Feature:
+        """
+        add a feature to annotation_db
+
+        Parameters
+        ----------
+        biotype
+            biological type
+        name
+            name of the feature
+        spans
+            coordinates for this sequence
+        parent_id
+            name of the feature parent
+        strand
+            '+' or '-', defaults to '+'
+        on_alignment
+            whether the feature spans are alignment coordinates
+        seqid
+            ignored since the feature is added to this sequence
+
+        Returns
+        -------
+        Feature instance
+        """
+        feature_data = FeatureDataType(
+            seqid=self.name,
+            **{n: v for n, v in locals().items() if n not in ("self", "seqid")},
+        )
+
+        self.annotation_db.add_feature(**feature_data)
+        for discard in ("on_alignment", "parent_id"):
+            feature_data.pop(discard)
+        return self.make_feature(feature_data)
+
+    def to_moltype(self, moltype):
```

Comment:
type hints

Reply:
that would have to be "Self" which is only supported in 3.11